feat: add LTI and OAuth2 integrations for authentication

Opeyem1a opened the pull request.

Opeyem1a committed:
feat: add LTI and OAuth2 integrations for authentication

Opeyem1a committed:
fix types

Opeyem1a committed:
lint

Opeyem1a committed:
fix types pt.2

Opeyem1a committed:
update macos runner

Opeyem1a committed:
try macos-14

Opeyem1a committed:
try binary

Opeyem1a committed:
try binary only

Opeyem1a committed:
conditional CI

catinhere commented on src/lti/views.py:
```diff
@@ -0,0 +1,143 @@
+    auth_user, created = MyUser.objects.get_or_create(
+        # Make the username unique to the launch ISS, so different
+        # institution's Canvas instance creates unique usernames
+        username=f"{canvas_user_id}+{launch_iss}"
```
`canvas_user_id` (and `canvas_course_id` used below) come from the LTI custom claims, which require manually adding a `custom_fields` block to the dev key. If it's missing, both values come back `None`, and every launch from that instance silently collides into a single `None+<iss>` user and a single `None`-keyed course instead of failing. Would be worth validating these claims are present and rejecting the launch if they're not

Opeyem1a replied:
Good catch, looks good

catinhere commented on src/lti/views.py:
```diff
@@ -0,0 +1,143 @@
+
+# do not require deployments in config
+class ExtendedDjangoMessageLaunch(DjangoMessageLaunch):
+    def validate_deployment(self):
```
`validate_deployment` is overridden to skip LTI's deployment id check. That check is an extra access-control layer on top of the issuer/signature validation to ensure only courses we've approved can use the tool. Since courses currently only get added manually after requesting access, this isn't urgent but it's worth considering.

Opeyem1a replied:
This was intentional for local development purposes, let me try to support both paths

catinhere committed:
fix: enforce LTI deployment validation and require custom claims

catinhere committed:
fix: enforce LTI deployment_id validation

Remove the ExtendedDjangoMessageLaunch override that was skipping validate_deployment(), so launches are checked against the deployment_ids actually registered in tool_conf
